File: packages/esm-format-go/pkg/esm/data_loader_fixtures_test.go
package esm

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

// TestDataLoaderFixturesCoverage verifies that the new STAC-like DataLoader
// schema can express every data loader currently implemented in EarthSciData.jl
// (gt-mos coverage acceptance test).
//
// For each EarthSciData.jl loader, a hand-authored fixture lives under
// testdata/data_loaders/<name>.esm. Each fixture must:
//  1. Schema-validate (gojsonschema + the embedded esm-schema.json).
//  2. Round-trip: parse -> serialize -> parse -> deep equal (as json.RawMessage).
//  3. Express the loader's parameter footprint from the EarthSciData.jl source.
func TestDataLoaderFixturesCoverage(t *testing.T) {
	// Each entry corresponds to an EarthSciData.jl loader. Add new entries here
	// when new loaders land in EarthSciData.jl; the fixture file pins the schema
	// coverage for that loader.
	cases := []struct {
		name     string
		fixture  string
		loaderID string
		kind     string
		minVars  int
	}{
		{"GEOSFP", "geosfp.esm", "GEOSFP_I3", "grid", 3},
		{"ERA5_PressureLevels", "era5.esm", "ERA5_PL", "grid", 7},
		{"WRF", "wrf.esm", "WRF_d01", "grid", 5},
		{"NEI2016Monthly", "nei2016monthly.esm", "NEI2016Monthly_ptegu", "grid", 4},
		{"CEDS", "ceds.esm", "CEDS_NOx", "grid", 3},
		{"EDGARv81Monthly", "edgar_v81_monthly.esm", "EDGAR_v81_Monthly_NOx_ENE", "grid", 1},
		{"USGS3DEP_Elevation", "usgs3dep.esm", "USGS3DEP_Elevation", "static", 1},
		{"USGS3DEP_Slopes", "usgs3dep_slopes.esm", "USGS3DEP_Slopes", "static", 2},
	}

	for _, tc := range cases {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			path := filepath.Join("testdata", "data_loaders", tc.fixture)
			data, err := os.ReadFile(path)
			if err != nil {
				t.Fatalf("read fixture %s: %v", path, err)
			}

			// 1. LoadString runs the embedded JSON schema check first.
			esmFile, err := LoadString(string(data))
			if err != nil {
				t.Fatalf("LoadString failed for %s: %v", tc.fixture, err)
			}

			loader, ok := esmFile.DataLoaders[tc.loaderID]
			if !ok {
				t.Fatalf("fixture %s missing expected data_loader %q", tc.fixture, tc.loaderID)
			}
			if loader.Kind != tc.kind {
				t.Errorf("loader %q kind: got %q, want %q", tc.loaderID, loader.Kind, tc.kind)
			}
			if loader.Source.URLTemplate == "" {
				t.Errorf("loader %q missing source.url_template", tc.loaderID)
			}
			if got := len(loader.Variables); got < tc.minVars {
				t.Errorf("loader %q: got %d variables, want >= %d", tc.loaderID, got, tc.minVars)
			}
			for varName, v := range loader.Variables {
				if v.FileVariable == "" {
					t.Errorf("loader %q variable %q missing file_variable", tc.loaderID, varName)
				}
				if v.Units == "" {
					t.Errorf("loader %q variable %q missing units", tc.loaderID, varName)
				}
			}

			// 2. Structural validation must pass (no errors for required fields).
			vres := Validate(esmFile)
			if !vres.Valid {
				t.Errorf("structural validation failed for %s: %+v", tc.fixture, vres.Messages)
			}

			// 3. Round-trip: serialize, re-parse, and compare the canonicalized
			// JSON trees. We compare via json.Unmarshal into generic interface{}
			// so that map key ordering and whitespace don't matter.
			serialized, err := Save(esmFile)
			if err != nil {
				t.Fatalf("Save failed for %s: %v", tc.fixture, err)
			}
			roundTripped, err := LoadString(serialized)
			if err != nil {
				t.Fatalf("LoadString(Save()) failed for %s: %v", tc.fixture, err)
			}
			origNorm, err := normalizeEsmJSON(esmFile)
			if err != nil {
				t.Fatalf("normalize original failed: %v", err)
			}
			rtNorm, err := normalizeEsmJSON(roundTripped)
			if err != nil {
				t.Fatalf("normalize round-tripped failed: %v", err)
			}
			if !reflect.DeepEqual(origNorm, rtNorm) {
				t.Errorf("round-trip produced different structure for %s", tc.fixture)
			}
		})
	}
}

// normalizeEsmJSON serializes an EsmFile and reparses it as interface{} so the
// result can be compared with reflect.DeepEqual without being affected by Go
// map ordering or float representation quirks.
func normalizeEsmJSON(f *EsmFile) (interface{}, error) {
	b, err := json.Marshal(f)
	if err != nil {
		return nil, err
	}
	var v interface{}
	if err := json.Unmarshal(b, &v); err != nil {
		return nil, err
	}
	return v, nil
}

File: packages/EarthSciSerialization.jl/test/data_loader_fixtures_test.jl
using Test
using EarthSciSerialization
using JSON3

# Verifies that the STAC-like DataLoader schema can express every
# EarthSciData.jl data source. For each fixture we:
#   1. Parse it.
#   2. Schema-validate against data/esm-schema.json.
#   3. Round-trip parse -> serialize -> parse and compare key fields.
#
# These fixtures are the concrete acceptance test for the gt-q4k schema
# redesign — if any EarthSciData loader cannot be expressed, this test
# suite fails and the gap should be escalated, not worked around.
@testset "DataLoader EarthSciData coverage fixtures" begin
    fixtures_dir = joinpath(@__DIR__, "fixtures", "data_loaders")
    @test isdir(fixtures_dir)

    fixture_files = sort(filter(f -> endswith(f, ".esm"),
                                 readdir(fixtures_dir)))
    @test !isempty(fixture_files)

    expected_loaders = [
        "geosfp.esm",
        "era5.esm",
        "wrf.esm",
        "nei2016monthly.esm",
        "ceds.esm",
        "edgar_v81.esm",
        "usgs3dep.esm",
    ]
    for name in expected_loaders
        @test name in fixture_files
    end

    for fname in fixture_files
        fpath = joinpath(fixtures_dir, fname)
        @testset "fixture $fname" begin
            # 1. Parse.
            original = EarthSciSerialization.load(fpath)
            @test original isa EarthSciSerialization.EsmFile
            @test original.data_loaders !== nothing
            @test !isempty(original.data_loaders)

            # 2. Schema-validate.
            result = EarthSciSerialization.validate(original)
            if !result.is_valid
                @info "Validation errors for $fname" errors=result.schema_errors structural=result.structural_errors
            end
            @test isempty(result.schema_errors)

            # 3. Round-trip.
            tmp = tempname() * ".esm"
            try
                EarthSciSerialization.save(tmp, original)
                reloaded = EarthSciSerialization.load(tmp)
                @test length(reloaded.data_loaders) == length(original.data_loaders)
                for (name, orig_loader) in original.data_loaders
                    @test haskey(reloaded.data_loaders, name)
                    reloaded_loader = reloaded.data_loaders[name]
                    @test reloaded_loader.kind == orig_loader.kind
                    @test reloaded_loader.source.url_template ==
                          orig_loader.source.url_template
                    @test keys(reloaded_loader.variables) ==
                          keys(orig_loader.variables)
                    for (var_name, orig_var) in orig_loader.variables
                        reloaded_var = reloaded_loader.variables[var_name]
                        @test reloaded_var.file_variable == orig_var.file_variable
                        @test reloaded_var.units == orig_var.units
                    end
                end
            finally
                isfile(tmp) && rm(tmp)
            end
        end
    end
end
